src/FixedPointNumbers.jl

```diff
@@ -106,36 +106,62 @@ floatmax(::Type{T}) where {T <: FixedPoint} = typemax(T)
 
 
 """
-    floattype(::Type{T})
+    floattype(::Type{T})::Type{<:AbstractFloat}
 
-Return the minimum float type that represents `T` without overflow to `Inf`.
+Return a minimal type suitable for performing computations with instances of type `T` without integer overflow.
 
-# Example
+The fallback definition of `floattype(T)` applies only to `T<:AbstractFloat`.
+However, it is permissible to extend `floattype` to return types that are not subtypes of
+`AbstractFloat`; the key characteristic is that the return type should support computation without integer overflow.
+
+In general the returned type should have the minimum bitwidth needed to encode the full precision of the input type.
+however, a priority should be placed on computational efficiency; consequently, types like `Float16` should be avoided
+except in scenarios where they are guaranteed to have hardware support.
+
+# Examples
 
 A classic usage is to avoid overflow behavior by promoting `FixedPoint` to `AbstractFloat`
 
-```julia
+```jldoctest
 julia> x = N0f8(1.0)
 1.0N0f8
 
 julia> x + x # overflow
 0.996N0f8
 
-julia> float_x = floattype(eltype(x))(x)
-1.0f0
+julia> T = floattype(x)
+Float32
 
-julia> float_x + float_x
+julia> T(x) + T(x)
 2.0f0
 ```
+
+The following represents a valid extension of `floattype` to non-AbstractFloats:
+
+```julia
+julia> using FixedPointNumbers, ColorTypes
+
+julia> floattype(RGB{N0f8})
+RGB{Float32}
+```
+
+`RGB` itself is not a subtype of `AbstractFloat`, but unlike `RGB{N0f8}` operations with `RGB{Float32}` are not subject to integer overflow.
 """
-floattype(::Type{T}) where {T <: Real} = T # fallback
+floattype(::Type{T}) where {T <: AbstractFloat} = T # fallback (we want a MethodError if no method producing AbstractFloat is defined)
 floattype(::Type{T}) where {T <: Union{ShortInts, Bool}} = Float32
 floattype(::Type{T}) where {T <: Integer} = Float64
 floattype(::Type{T}) where {T <: LongInts} = BigFloat
 floattype(::Type{X}) where {T <: ShortInts, X <: FixedPoint{T}} = Float32
 floattype(::Type{X}) where {T <: Integer, X <: FixedPoint{T}} = Float64
+floattype(::Type{X}) where {T <: Integer, X <: Rational{T}} = typeof(zero(T)/oneunit(T))
 floattype(::Type{X}) where {T <: LongInts, X <: FixedPoint{T}} = BigFloat
 
+# Non-Real types
+floattype(::Type{Complex{T}}) where T = Complex{floattype(T)}
+floattype(::Type{<:Irrational}) = Float64
```

Comment:
```suggestion
floattype(::Type{<:AbstractIrrational}) = Float64
```
This is not clear, but in terms of precision, `Irrational` and other subtypes of `AbstractIrrational` should have the same result.

BTW, `AbstractIrrational` is a `Real` type. (see the caption comment "# Non-Real types")